Gazebo new vehicles, ROS 2 frame conventions, multi vehicle simulation

beniaminopozzan opened the pull request:
This PR covers four main topics

## Update Gazebo docs with new vehicles support

- PX4 now supports fixed wings and standard VTOL + x500 with visual odometry and x500 with deep camera
- `make px4_sitl list_vmd_make_targets` has been removed (it does not exist anymore)

## Update xrce-dds and ROS 2 docs with more details about custom topics and message support and namespaces

- Deeper explanation of the role of `dds_topics.yaml` is provided.
- Instructions for manually synchronizing `PX4-Autopilot/msg` and `px4_msgs/msg` are provided.
- How to define custom namespaces, both in real vehicles and in simulations.

The same topics are covered in both xrce-dds page and ROS 2, with the exact same words, maybe it is better to just have one linking to the other.

## On the ROS 2 page, compatibility note about the different frame conventions is added

This is the first step toward a better rewriting of the documentation about external vision odometry https://docs.px4.io/main/en/tutorials/motion-capture.html and https://docs.px4.io/main/en/ros/external_position_estimation.html

## ROS 2 multi vehicle instructions

- A new "multi vehicle" page is added in the ROS 2 section to cover the parts that are common to all simulators
  - unique session key
  - unique namespace
  - the issue with `target_system`
- The "multi vehicle simulations" pages in simulation are updated.
  - The general one now mentions that Gazebo supports ROS. Moreover, Gazebo and Gazebo Classic allow for different kind of vehicle to be simulated together.
  - For Gazebo Classic all paths and instructions are updated
  - For Gazebo new instructions are provided for interfacing with ROS 2
- Continues #2345

---

Regarding the micrortps cleanup, only the page about ROS 1 over ROS 2 remains.

hamishwillee commented:
@beniaminopozzan 

> make px4_sitl list_vmd_make_targets

I know that this was removed, but never had an explanation of why. It was so very useful. Do you know/can we have it back?

beniaminopozzan commented:
@hamishwillee 
My understanding is that it was removed because now the simulation targets are defined in multiple CMakeLists, one for each simulator. I guess it can be recovered and splitted. In the end it will be something like
```sh
make px4_sitl list_vmd_make_gz_targets
make px4_sitl list_vmd_make_gazebo-classic_targets
make px4_sitl list_vmd_make_flightgear_targets
```

hamishwillee commented on en/middleware/xrce_dds.md:
```diff
@@ -233,6 +233,9 @@ At time of writing there is no PX4 parameter for setting the address of the remo
 Therefore for Ethernet connections you will _have_ to use a command to start the client.
 :::
 
+On the simulator the client is automatically started on localhost UDP port 8888.
+Moreover, the startup scripts take into account [multi vehicle simulations](../ros/ros2_multi_vehicle.md) and their configuration.
```
FYI these bits might lead to some conflict with #2352

hamishwillee replied:
@beniaminopozzan The second line confuses me a bit. We start using parameters most of the time "in future" so is this saying that parameters won't take into account multi vehicle, or that if I write a start up script it has to take into account multi-vehicle simulation?

beniaminopozzan replied:
@hamishwillee Yes, most likely they will

At the moment you can do almost everything with parameters: set the IP, port, session key. However, namespaces cannot be set with parameters and users need to start the agent manually and use either the `-s` argument or `PX4_MICRODDS_NS`. There is no other way.
To have multi vehicle setups (simulation or real vehicles) you need to set namespaces.
Therefore:

- Parameters won't take into account multi vehicle? Yes, they won't. And this behavior will not change until string parameter will be supported in PX4.
- If I write a start up script it has to take into account multi-vehicle simulation? If you write a startup script and you want to perform multi-vehicle simulation, then yes, your script must handle them. If you don't want to perform multi-vehicle simulation then no, your script doesn't have to handle them.

hamishwillee replied:
Thanks for the clarification. I guess it is a bit confusing because we're talking about "the startup scripts" here as though users know what specific scripts you mean.

How about this?

```suggestion
On the simulator the client is automatically started on localhost UDP port 8888.

Note that startup commands are used in startup scripts for [multi vehicle simulations](../ros/ros2_multi_vehicle.md), enabling the setting of appropriate instance ids and DDS namespace.
```

beniaminopozzan replied:
As

> On the simulator the client is automatically started on localhost UDP port 8888.

is already mentioned in the first note for "Starting the Client", another possibility is

```suggestion
The simulator [startup logic](../concept/system_startup.md) ([init.d-posix/rcS](https://github.com/PX4/PX4-Autopilot/blob/main/ROMFS/px4fmu_common/init.d-posix/rcS)) uses the client startup commands for single and [multi vehicle simulations](../ros/ros2_multi_vehicle.md), enabling the setting of appropriate instance ids and DDS namespaces.
```

hamishwillee replied:
Yes.

hamishwillee commented on en/middleware/xrce_dds.md:
```diff
@@ -241,23 +244,90 @@ The topics are release specific (support is compiled into [microdds-client](../m
 
-Note that ROS 2/DDS needs to have the _same_ message definitions that were used to create the XRCE-DDS client module in the PX4 Firmware in order to interpret the messages:
+Note that ROS 2/DDS needs to have the _same_ message definitions that were used to create the XRCE-DDS client module in the PX4 Firmware in order to interpret the messages.
+The definition are stored in the ROS 2 interface package [PX4/px4_msgs](https://github.com/PX4/px4_msgs) and they are automatically synchronized by CI on the `main` and release branches.
```
@beniaminopozzan It used to be that they included all messages - do they now just contain the ones that are available?

beniaminopozzan replied:
@hamishwillee it still includes all messages as it generally simplify the setup. Why? Because if
all messages are included then a user who just wants to add a new topic to the bridge only needs to update dds_topics.yaml without even recompiling the ROS 2 workspace.

But this is open to discussion. For example a good reason to limit the messages copied into px4_msgs only to those in dds_topics.yaml is that there are *a lot* of messages, and the compilation time can be quite annoying. Even more if you are compiling on a Raspberry Pi.

hamishwillee replied:
Fair enough. I think it worth clarifying that just because a message is in that folder doesn't mean it is built.

```suggestion
All PX4 message definitions are stored in the ROS 2 interface package [PX4/px4_msgs](https://github.com/PX4/px4_msgs), and they are automatically synchronized by CI on the `main` and release branches.
```

hamishwillee commented on en/middleware/xrce_dds.md:
```diff
@@ -241,23 +244,90 @@ The topics are release specific (support is compiled into [microdds-client](../m
-- If you're creating your own new topic you will need to update [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml) and also copy the message definitions out of PX4 source tree into any ROS 2 workspace that needs them.
-  The messages must be in a folder named `px4_msgs/msg/`.
+- If you're using a main or release version of PX4 you can get the message definitions by cloning the interface package [PX4/px4_msgs](https://github.com/PX4/px4_msgs) into your workspace.
+- If you're creating your own new topics e/o uORB messages you will need to update [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml), clone the the interface package and then manually synchronize it by copying the new/modified message definitions from [PX4-Autopilot/msg](https://github.com/PX4/PX4-Autopilot/tree/main/msg) to its `msg` fo ...
```
This is just to add a line break:
```suggestion
- If you're creating your own new topics e/o uORB messages you will need to update [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml), clone the the interface package and then manually synchronize it by copying the new/modified message definitions from [PX4-Autopilot/msg](https://github.com/PX4/PX4-Autopilot/tree/main/msg) to its `msg` folders. Assuming that PX4-Autopilot is in your home directory `~` while `px4_msgs` is in `~/px4_ros_com/src/`, then the command might be:

```

hamishwillee commented on en/middleware/xrce_dds.md:
```diff
@@ -241,23 +244,90 @@ The topics are release specific (support is compiled into [microdds-client](../m
+  
+  Each (`topic`,`type`) pairs defines
+
+  1. A new subscription or publication depending on the list it is added.
```
```suggestion
  1. A new subscription or publication depending on the list to which it is added.
```

hamishwillee commented on en/middleware/xrce_dds.md:
```diff
@@ -241,23 +244,90 @@ The topics are release specific (support is compiled into [microdds-client](../m
+  Each (`topic`,`type`) pairs defines
+
+  1. A new subscription or publication depending on the list it is added.
+  2. The topic _base name_, which MUST coincide with the desired uORB topic name that you want to publish/subscribe.
```
FYI There is nothing wrong with highlighting MUST, but I prefer to use bold/italic, and have done so across the docs. There are exceptions.

```suggestion
  2. The topic _base name_, which **must** coincide with the desired uORB topic name that you want to publish/subscribe.
```

hamishwillee commented on en/middleware/xrce_dds.md:
```diff
@@ -241,23 +244,90 @@ The topics are release specific (support is compiled into [microdds-client](../m
+     It is identified by the last token in `topic:` that starts with `/` and does not contains any `/` in it.
+     `vehicle_odometry`, `vehicle_status` and `offboard_control_mode` are examples of base names.
+  3. The topic [namespace](https://design.ros2.org/articles/topic_and_service_names.html#namespaces). 
+     By default it is set to
+      - `/fmu/out/` for topics that are _published_ by PX4.
```
FYI, Sentence preceding list should start with `:`. Empty line of space also nice.
```suggestion
     By default it is set to:
     
      - `/fmu/out/` for topics that are _published_ by PX4.
```

hamishwillee commented on en/middleware/xrce_dds.md:
```diff
@@ -241,23 +244,90 @@ The topics are release specific (support is compiled into [microdds-client](../m
+  3. The topic [namespace](https://design.ros2.org/articles/topic_and_service_names.html#namespaces). 
+     By default it is set to
+      - `/fmu/out/` for topics that are _published_ by PX4.
+      - `/fmu/in/` for topics that are _subscirbed_ by PX4.
```
```suggestion
      - `/fmu/in/` for topics that are _subscribed_ by PX4.
```

hamishwillee commented on en/middleware/xrce_dds.md:
```diff
@@ -241,23 +244,90 @@ The topics are release specific (support is compiled into [microdds-client](../m
+     By default it is set to
+      - `/fmu/out/` for topics that are _published_ by PX4.
+      - `/fmu/in/` for topics that are _subscirbed_ by PX4.
+  4. The message type (`VehicleOdometry`, `VehicleStatus`, `OffboardControlMode`, ecc) and the ROS 2 package (`px4_msgs`) that is expected to provide the message definition.
```
```suggestion
  4. The message type (`VehicleOdometry`, `VehicleStatus`, `OffboardControlMode`, etc.) and the ROS 2 package (`px4_msgs`) that is expected to provide the message definition.
```

hamishwillee commented on en/middleware/xrce_dds.md:
```diff
@@ -241,23 +244,90 @@ The topics are release specific (support is compiled into [microdds-client](../m
+      - `/fmu/in/` for topics that are _subscirbed_ by PX4.
+  4. The message type (`VehicleOdometry`, `VehicleStatus`, `OffboardControlMode`, ecc) and the ROS 2 package (`px4_msgs`) that is expected to provide the message definition.
+
+  You can arbitrary change the configuration.
```
```suggestion
  You can arbitrarily change the configuration.
```

hamishwillee commented on en/middleware/xrce_dds.md:
```diff
@@ -241,23 +244,90 @@ The topics are release specific (support is compiled into [microdds-client](../m
 
+### Customizing the topic namespace
+
+Custom topic namespaces can be applied at build time (changing [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml)) or at runtime (useful for multi vehicle operations):
```
```suggestion
Custom topic namespaces can be applied at build time (changing [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml)) or at runtime (which is useful for multi vehicle operations):
```

hamishwillee commented on en/middleware/xrce_dds.md:
```diff
@@ -241,23 +244,90 @@ The topics are release specific (support is compiled into [microdds-client](../m
+
+Custom topic namespaces can be applied at build time (changing [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml)) or at runtime (useful for multi vehicle operations):
+
+ - One possibility is to use the `-n` option when starting the [microdds-client](../modules/modules_system.md#microdds-client) from command line. This technique can be used both in simulation and real vehicles.
```
FYI only. I prefer to break on sentences. That is to my taste only.

```suggestion
 - One possibility is to use the `-n` option when starting the [microdds-client](../modules/modules_system.md#microdds-client) from command line.
   This technique can be used both in simulation and real vehicles.
```

hamishwillee commented on en/middleware/xrce_dds.md:
```diff
@@ -241,23 +244,90 @@ The topics are release specific (support is compiled into [microdds-client](../m
+Custom topic namespaces can be applied at build time (changing [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml)) or at runtime (useful for multi vehicle operations):
+
+ - One possibility is to use the `-n` option when starting the [microdds-client](../modules/modules_system.md#microdds-client) from command line. This technique can be used both in simulation and real vehicles.
+ - Only in simulation, where the client is automatically started, a custom namespace can be provided setting the environment variable `PX4_MICRODDS_NS` before starting the simulation.
```
```suggestion
 - a custom namespace can be provided for simulations (only) by setting the environment variable `PX4_MICRODDS_NS` before starting the simulation.
```

beniaminopozzan replied:
LGTM @hamishwillee

hamishwillee commented on en/middleware/xrce_dds.md:
```diff
@@ -241,23 +244,90 @@ The topics are release specific (support is compiled into [microdds-client](../m
+
+:::note
+Changing the namespace at runtime will append the desired namespace as a prefix to all `topic` fields in [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml).
+Therefore, commands like
```
```suggestion
Therefore, commands like:

```

hamishwillee commented on en/middleware/xrce_dds.md:
````diff
@@ -241,23 +244,90 @@ The topics are release specific (support is compiled into [microdds-client](../m
+```sh
+microdds_client start -n uav_1
+```
+or
````
```suggestion

or

```

hamishwillee commented on en/middleware/xrce_dds.md:
````diff
@@ -241,23 +244,90 @@ The topics are release specific (support is compiled into [microdds-client](../m
+```sh
+PX4_MICRODDS_NS=uav_1 make px4_sitl gz_x500
+```
+will generate topics under the namespaces
````
```suggestion

will generate topics under the namespaces:

```

hamishwillee commented on en/ros/ros2_multi_vehicle.md:
```diff
@@ -0,0 +1,54 @@
+
+## Principle of operation
+
+In simulation each PX4 instance receives an unique `px4_instance` number starting from `0`.
```
```suggestion
In simulation each PX4 instance receives a unique `px4_instance` number starting from `0`.
```

hamishwillee commented on en/ros/ros2_multi_vehicle.md:
````diff
@@ -0,0 +1,54 @@
+```
+
+:::note
+By doing so, `XRCE_DDS_KEY` will always coincide with [MAV_COMP_ID](../advanced_config/parameter_reference.md#MAV_COMP_ID).
````
Do you mean MAV_SYS_ID ?

beniaminopozzan replied:
Yes, sorry

hamishwillee commented on en/ros/ros2_multi_vehicle.md:
````diff
@@ -0,0 +1,54 @@
+```
+
+:::note
+The environment variable `PX4_MICRODDS_NS`, if set, will override the above mentioned namespace behavior.
````
```suggestion
The environment variable `PX4_MICRODDS_NS`, if set, will override the namespace behavior described above.
```

hamishwillee commented on en/ros/ros2_multi_vehicle.md:
```diff
@@ -0,0 +1,54 @@
+The environment variable `PX4_MICRODDS_NS`, if set, will override the above mentioned namespace behavior.
+:::
+
+The first instance (`px4_instance=0`) does not have an additional namespace in order to be consistent with the default behavior of the xrce-dds client on real vehicle.
```
```suggestion
The first instance (`px4_instance=0`) does not have an additional namespace in order to be consistent with the default behavior of the `xrce-dds` client on a real vehicle.
```

hamishwillee commented on en/ros/ros2_multi_vehicle.md:
```diff
@@ -0,0 +1,54 @@
+The first instance (`px4_instance=0`) does not have an additional namespace in order to be consistent with the default behavior of the xrce-dds client on real vehicle.
+This mismatch can be fixed by manually using `PX4_MICRODDS_NS` on the first instance or by starting adding vehicles from index `1` instead of `0` (this is the default behavior adopted by [sitl_multiple_run.sh](https://github.com/PX4/PX4-Autopilot/blob/main/Tools/simulation/gazebo-classic/sitl_multiple_run.sh) for Gazebo Classic).
+
+The default client configuration in simulation is summarized as follows
```
```suggestion
The default client configuration in simulation is summarized as follows:
```

hamishwillee commented on en/ros/ros2_multi_vehicle.md:
```diff
@@ -0,0 +1,54 @@
+
+The default client configuration in simulation is summarized as follows
+
+| `PX4_MICRODDS_NS` | `px4_instance` | `XRCE_DDS_KEY`   | client namespace      |
```
Should this be `px4_instance` as first column?

beniaminopozzan replied:
Yes, it could be. I did not think about the order of the two. You proposed that because `px4_instance` is sort of "more important"?

hamishwillee replied:
No, I proposed it because it seems to me that the instance is what decides the namespace and key.  If you have a single instance setup you don't care about the other values, so it is what generally determines if the others are relevant. Make sense?

beniaminopozzan replied:
If I have a single instance setup I may still want to add a custom namespace with `PX4_MICRODDS_NS`. But indeed `px4_instance` is the only one that affects both `XRCE_DDS_KEY`and the final client namespace.

hamishwillee replied:
UP to you! I'm going to leave this for you now.

beniaminopozzan replied:
I'm going to leave it as it is then.

hamishwillee commented on en/ros/ros2_multi_vehicle.md:
```diff
@@ -0,0 +1,54 @@
+
+## Adjusting the `target_system` value
+
+PX4 accepts [VehicleCommand](../msg_docs/VehicleCommand.md) messages only if their `target_system` field is zero (broadcast communication) or coincides with `MAV_COMP_ID`.
```
Again, probably MAV_SYS_ID . Sys ID is used for mavlink "vehicle level" addressing. Comp id is used for addressing something like a companion computer or a mavlink camera.

hamishwillee commented on en/sim_gazebo_classic/gazebo_vehicles.md:
```diff
@@ -4,10 +4,6 @@ This topic lists/displays the vehicles supported by the PX4 [Gazebo Classic](../
 Supported vehicle types include: mutirotors, VTOL, VTOL Tailsitter, Plane, Rover, Submarine/UUV.
 
-:::tip
-For the full list of build targets run `make px4_sitl list_vmd_make_targets` (and filter on those that start with `gazebo_`).
```
Fair enough, but I want this feature back :-(. So useful.

hamishwillee commented on en/sim_gazebo_classic/multi_vehicle_simulation_gazebo.md:
```diff
@@ -55,21 +55,13 @@ The `MAV_SYS_ID` and various UDP ports are allocated in the SITL rcS: [init.d-po
 
 <a id="with_dds"></a>
-### Build and Test (RTPS/DDS)
+### Build and Test (xrce-dds)
```
Why lower case?

```suggestion
### Build and Test (XRCE-DDS)
```

beniaminopozzan replied:
My mistake

hamishwillee commented on en/sim_gazebo_classic/multi_vehicle_simulation_gazebo.md:
```diff
@@ -55,21 +55,13 @@ The `MAV_SYS_ID` and various UDP ports are allocated in the SITL rcS: [init.d-po
 :::note
-You will need to have installed or *eProsima Fast DDS* or ROS 2 Foxy or above and the `micrortps_agent` should be run in the different terminals for each vehicle.
-For more information see: [RTPS/DDS Interface: PX4-Fast RTPS(DDS) Bridge](../middleware/micrortps.md), for how to use the interaction with non-ROS 2 DDS participant applications, or [ROS 2 User Guide (PX4-ROS 2 Bridge)](../ros/ros2_comm.md), for interfacing with ROS 2 nodes.
+You will need to have installed the xrce-dds dependencies.
```
```suggestion
You will need to have installed the XRCE-DDS dependencies.
```

hamishwillee commented on en/sim_gazebo_classic/multi_vehicle_simulation_gazebo.md:
```diff
@@ -78,41 +70,33 @@ To build an example setup, follow the steps below:
 
-1. Run `micrortps_agent`.
-   For example, to connect 4 vehicles, run:
+2. Run `micro xrce-dds agent`. It will connect automatically to all four vehicles
```
```suggestion
2. Run `MicroXRCEAgent`.
   It will automatically connect to all four vehicles:
```

hamishwillee commented on en/sim_gazebo_classic/multi_vehicle_simulation_gazebo.md:
```diff
@@ -78,41 +70,33 @@ To build an example setup, follow the steps below:
-   In order to communicate with a specific instance of PX4 using ROS 2, you must use the `-n <namespace>` option.
-   For example, running `micrortps_agent -t UDP -r 2020 -s 2019 -n vhcl0` will result in the agent publishing all its topics with the namespace prefix `/vhcl0` (eg. `SensorCombined` data from `vhcl0` will be published on the topic `/vhcl0/fmu/sensor_combined/out`, while if one wants to send commands to the same vehicle, it has to publish to topic `/vhcl0/fmu/vehicle_command/in`).
-   You can then subscribe and publish to just that vehicle's topics.
+   The simulator startup script automatically assign [unique namespaces](../ros/ros2_multi_vehicle.md) to each vehicle.
```
```suggestion
   The simulator startup script automatically assigns a [unique namespace](../ros/ros2_multi_vehicle.md) to each vehicle.
```

hamishwillee commented on en/sim_gazebo_gz/gazebo_vehicles.md:
```diff
@@ -0,0 +1,62 @@
+# Gazebo Vehicles
+
+This topic lists/displays the vehicles supported by the PX4 [Gazebo](../sim_gazebo_gz/README.md) simulation and the `make` commands required to run them (the commands are run from a terminal in the **PX4-Autopilot** directory).
```
```suggestion
This topic lists/displays the vehicles supported by the PX4 [Gazebo](../sim_gazebo_gz/README.md) simulation, and the `make` commands required to run them (the commands are run from a terminal in the **PX4-Autopilot** directory).
```

hamishwillee commented on en/sim_gazebo_gz/gazebo_vehicles.md:
```diff
@@ -0,0 +1,62 @@
+The [Gazebo](../sim_gazebo_gz/README.md) page shows how to install Gazebo.
+:::
+
+## Multicopter
+<a id="x500-quadrotor"></a>
+### x500 quadrotor
```
Note, headings use first letter capitalization of all words (except maybe 'a', 'of' etc). You only need an anchor if you can't use the default on.

```suggestion
## Multicopter

### X500 Quadrotor
```

hamishwillee commented on en/sim_gazebo_gz/gazebo_vehicles.md:
````diff
@@ -0,0 +1,62 @@
+make px4_sitl gz_x500
+```
+
+<a id="x500-quadrotor-with-visual-odometry"></a>
+### x500 quadrotor with visual odometry
````
```suggestion
### X500 Quadrotor with Visual Odometry
```

hamishwillee commented on en/sim_gazebo_gz/gazebo_vehicles.md:
```diff
@@ -0,0 +1,62 @@
+
+![x500 in Gazebo](../../assets/simulation/gazebo/vehicles/x500.png)
+
+<a id="x500-quadrotor-with-depth-camera"></a>
+### x500 quadrotor with Depth Camera
```
```suggestion
### X500 Quadrotor with Depth Camera
```

hamishwillee commented on en/sim_gazebo_gz/gazebo_vehicles.md:
```diff
@@ -0,0 +1,60 @@
+
+## VTOL
+
+<a id="standard_vtol"></a>
+### Standard VTOL
```
Not needed - you link to this correctly above.

```suggestion
### Standard VTOL
```

hamishwillee commented on en/sim_gazebo_gz/gazebo_vehicles.md:
```diff
@@ -0,0 +1,60 @@
+<a id="fixed_wing"></a>
+## Plane/Fixed Wing
+
+<a id="rc-cessna"></a>
```
FYI this is correct, though you could link to standard-plane if you wanted

hamishwillee replied:
Links are "better" if you use the Vupress heading anchor - that always takes you to the right place. Manual anchors like this are always a little off where you want them to go. Sometimes I have used manual anchors because they stay even if the heading name changes - things don't get broken unexpectedly. Also if headings are very long so the link is too.

Anyway, this anchor is fine.

hamishwillee commented on en/sim_gazebo_gz/multi_vehicle_simulation.md:
```diff
@@ -27,3 +27,101 @@ ARGS ./build/px4_sitl_default/bin/px4 [-i <instance>]
+- First follow the installation instructions for [Gazebo](../sim_gazebo_gz/README.md).
+- Then configure your system for [ROS 2 / PX4 operations](../ros/ros2_comm.md#installation-setup).
+- In different terminals manually start a multi vehicle simulation.
+  This example spawns 2 x500 quadrotors and a fixed wing
```
```suggestion
  This example spawns 2 X500 Quadrotors and a fixed wing.
```

hamishwillee commented on en/sim_gazebo_gz/multi_vehicle_simulation.md:
````diff
@@ -27,3 +27,101 @@ ARGS ./build/px4_sitl_default/bin/px4 [-i <instance>]
+  ```sh
+  MicroXRCEAgent udp4 -p 8888
+  ```
+- The agent will automatically connect to all clients and running `ros2 topic list` will return
````
```suggestion
  The agent will automatically connect to all clients.
- Run `ros2 topic list`. This should return:
```

hamishwillee commented on en/sim_gazebo_gz/multi_vehicle_simulation.md:
````diff
@@ -27,3 +27,107 @@ ARGS ./build/px4_sitl_default/bin/px4 [-i <instance>]
+  ```sh
+  MicroXRCEAgent udp4 -p 8888
+  ```
+- The agent will automatically connect to all clients and running `ros2 topic list` will return:
+- 
````
```suggestion
  The agent will automatically connect to all clients.
-  Run `ros2 topic list` to see the topic list, which should look like this:
```

hamishwillee commented on en/simulation/README.md:
```diff
@@ -37,7 +37,7 @@ It is not required to *use* the simulators.
 ## Simulator MAVLink API
 
-All simulators communicate with PX4 using the Simulator MAVLink API.
+All simulators but Gazebo communicate with PX4 using the Simulator MAVLink API.
```
```suggestion
All simulators except for Gazebo communicate with PX4 using the Simulator MAVLink API.
```

hamishwillee commented on en/simulation/README.md:
```diff
@@ -61,6 +61,7 @@ Message | Direction | Description
 [HIL_STATE_QUATERNION](https://mavlink.io/en/messages/common.html#HIL_STATE_QUATERNION) | Sim to PX4 | Contains the actual "simulated" vehicle position, attitude, speed etc. This can be logged and compared to PX4's estimates for analysis and debugging (for example, checking how well an estimator works for noisy (simulated) sensor inputs).
 [HIL_RC_INPUTS_RAW](https://mavlink.io/en/messages/common.html#HIL_RC_INPUTS_RAW) | Sim to PX4 | The RAW values of the RC channels received.
 
+PX4 directly uses [Gazebo API](https://gazebosim.org/docs) to interface with [Gazebo](../sim_gazebo_gz/README.md) and MAVlink is not required.
```
```suggestion
PX4 directly usesthe  [Gazebo API](https://gazebosim.org/docs) to interface with [Gazebo](../sim_gazebo_gz/README.md) and MAVlink is not required.
```

hamishwillee commented on en/simulation/README.md:
```diff
@@ -89,7 +90,7 @@ See [System Startup](../concept/system_startup.md) to learn more.
 ## SITL Simulation Environment
 
-The diagram below shows a typical SITL simulation environment for any of the supported simulators.
+The diagram below shows a typical SITL simulation environment for any of the supported simulators but Gazebo.
```
```suggestion
The diagram below shows a typical SITL simulation environment for any of the supported simulators that use MAVLink (i.e. all of them except Gazebo).
```

hamishwillee commented on en/middleware/xrce_dds.md:
````diff
@@ -241,23 +244,97 @@ The topics are release specific (support is compiled into [microdds-client](../m
-  ```yaml
-  - topic: /fmu/out/vehicle_odometry
-    type: px4_msgs::msg::VehicleOdometry
+  Technically, [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml) completely defines the pairing between PX4 message and topics and the ROS 2 ones.
````
Below seems like a bit of duplication with the ROS2 additions for custom uorb. Propose we add a section

"DDS Topics YAML" which covers the format and is linked from both these places?

hamishwillee commented on en/middleware/xrce_dds.md:
```diff
@@ -241,23 +244,97 @@ The topics are release specific (support is compiled into [microdds-client](../m
   :::
 
 
+### Customizing the topic namespace
```
This is a lower level heading (3) than the parent (2 - ""Supported uORB Messages"). Is that the intent? If not, perhaps this should go up one level.

hamishwillee commented on en/middleware/xrce_dds.md:
````diff
@@ -241,23 +244,97 @@ The topics are release specific (support is compiled into [microdds-client](../m
+```
+:::
+
 ### PX4 ROS 2 QoS Settings
````
This heading is nested below Supported "uORB Messages" too. Probably should go up one level as well.

hamishwillee commented on en/ros/ros2_comm.md:
````diff
@@ -442,23 +483,93 @@ For more information see [Starting XRCE-DDS](../middleware/xrce_dds.md#starting-
-  ```yaml
-  - topic: /fmu/out/vehicle_odometry
-    type: px4_msgs::msg::VehicleOdometry
+  Technically, [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml) completely defines the pairing between PX4 message and topics and the ROS 2 ones.
````
See comments above related to perhaps pushing this into its own section in the XRCE-DDS doc. Or here, but just once.

hamishwillee commented:
@beniaminopozzan Wow. That was very hard to review; my only consolation is that I expect it was much harder to write, and you have done quite an amazing about of work and testing. Thank you.

I might well have missed some things, but I think this is more than good enough. Some comments inline for you though.

hamishwillee commented:
> @hamishwillee My understanding is that it was removed because now the simulation targets are defined in multiple CMakeLists, one for each simulator. I guess it can be recovered and split. In the end it will be something like
> 
> ```shell
> make px4_sitl list_vmd_make_gz_targets
> make px4_sitl list_vmd_make_gazebo-classic_targets
> make px4_sitl list_vmd_make_flightgear_targets
> ```

Thanks for the explanation. One command would be better, but any way to understand your options would be useful.

beniaminopozzan commented:
> @beniaminopozzan Wow. That was very hard to review; my only consolation is that I expect it was much harder to write, and you have done quite an amazing about of work and testing. Thank you.
> 
> I might well have missed some things, but I think this is more than good enough. Some comments inline for you though.

I was quite proud of the PR until I saw almost 60 comments on it. Next time I'll spend much more time on checking and double checking for typos and on refining my English. I learned a lot going through your review today, thanks @hamishwillee !

hamishwillee commented:
> I was quite proud of the PR until I saw almost 60 comments on it. Next time I'll spend much more time on checking and double checking for typos and on refining my English. I learned a lot going through your review today, thanks @hamishwillee !

You should be proud of this PR. If you're going to contribute something this big you're going to make mistakes. Many of my "corrections" are only to my taste anyway.

That said, where possible it is often better to split up PRs into smaller chunks to ease review. You might, for example, have done the frame conventions stuff  in a separate PR. Which reminds me, we should go to the places that rely on these and link to that section instead of duplicating once this is in.

beniaminopozzan commented:
> > I was quite proud of the PR until I saw almost 60 comments on it. Next time I'll spend much more time on checking and double checking for typos and on refining my English. I learned a lot going through your review today, thanks @hamishwillee !
> 
> You should be proud of this PR. If you're going to contribute something this big you're going to make mistakes. Many of my "corrections" are only to my taste anyway.
> 
> That said, where possible it is often better to split up PRs into smaller chunks to ease review. You might, for example, have done the frame conventions stuff in a separate PR. Which reminds me, we should go to the places that rely on these and link to that section instead of duplicating once this is in.

I'll definitively make smaller PR next time!

I choose to duplicate the frame conventions part here because all other places where it mentioned are a bit out of date. I thought it will be easier to update them having the "right reference" already in the right place.

hamishwillee commented on en/middleware/xrce_dds.md:
```diff
@@ -306,6 +344,59 @@ ROS 2 uses the following QoS settings (by default) for publishers and subscripti
 <!-- From https://github.com/PX4/PX4-user_guide/pull/2259#discussion_r1099788316 -->
 
+
+## DDS Topics YAML
```
@beniaminopozzan OK, I moved both headings up one level, and added this heading here for DDS Topics YAML. I have linked to this section (which duplicates your original text) from above and from the ROS2 Guide). We might even have more information/links here, but this is enough for now - i.e. in some places we talk a little about the file format where we could redirect here.

Before I did all that I rebased on main, so this now includes all your other PRs.

I'm going to leave with you to think about this as I have to go catch a plane. If you want to do a further round of edits let me know, otherwise we can merge and add next bits as a post-process. UP to you. I'll look in to see if you have commented in a couple of hours.

beniaminopozzan commented on en/middleware/xrce_dds.md:
```diff
@@ -257,24 +260,59 @@ The topics are release specific (support is compiled into [microdds-client](../m
+
+- If you're using a main or release version of PX4 you can get the message definitions by cloning the interface package [PX4/px4_msgs](https://github.com/PX4/px4_msgs) into your workspace.
+- If you're creating or modifying uORB messages you must manually update the messages in your workspace from your PX4 source tree.
+  Generally this means that you would update [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml), clone the the interface package, and then manually synchronize it by copying the new/modified message definitions from [PX4-Autopilot/msg](https://github.com/PX4/PX4-Autopilot/tree/main/msg) to its `msg` folders.
```
```suggestion
  Generally this means that you would update [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml), clone the interface package, and then manually synchronize it by copying the new/modified message definitions from [PX4-Autopilot/msg](https://github.com/PX4/PX4-Autopilot/tree/main/msg) to its `msg` folders.
```

beniaminopozzan commented on en/ros/ros2_comm.md:
```diff
@@ -281,6 +281,47 @@ Note that ROS code does not have to set QoS settings when publishing (the PX4 se
+- To rotate a vector from FLU to FRD a pi rotation around the `X`-axis (front) is sufficient.
+- To rotate a vector from FRD to FLU a pi rotation around the `X`-axis (front) is sufficient.
+
+Examples of vectors that require rotation are
```
```suggestion
Examples of vectors that require rotation are:
```

beniaminopozzan commented on en/ros/ros2_comm.md:
```diff
@@ -448,23 +489,51 @@ For more information see [Starting XRCE-DDS](../middleware/xrce_dds.md#starting-
-ROS 2 needs to have the _same_ message definitions that were used to create the XRCE-DDS client module in the PX4 Firmware.
-These are automatically exported to [PX4/px4_msgs](https://github.com/PX4/px4_msgs) for main and release builds, and you can just clone them into your workspace.
+ROS 2 needs to have the _same_ message definitions that were used to create the XRCE-DDS client module in the PX4 Firmware in order to interpret the messages.
+The definition are stored in the ROS 2 interface package [PX4/px4_msgs](https://github.com/PX4/px4_msgs) and they are automatically synchronized by CI on the `main` and release branches.
```
```suggestion
The definition are stored in the ROS 2 interface package [PX4/px4_msgs](https://github.com/PX4/px4_msgs) and they are automatically synchronized by CI on the `main` and release branches.
Note that all the messages from PX4 source code are present in the repository, but only those listed in `dds_topics.yaml` will be available as ROS 2 topics.
```

beniaminopozzan commented on en/ros/ros2_comm.md:
```diff
@@ -448,23 +489,51 @@ For more information see [Starting XRCE-DDS](../middleware/xrce_dds.md#starting-
-- Copy the msg definitions out of the PX4 source tree into any ROS 2 workspace that needs them.
-  The messages must be in a folder named `px4_msgs/msg/`.
+- If you're using a main or release version of PX4 you can get the message definitions by cloning the interface package [PX4/px4_msgs](https://github.com/PX4/px4_msgs) into your workspace.
+- If you're creating your own new topics e/o uORB messages you will need to update [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml), clone the the interface package and then manually synchronize it by copying the new/modified message definitions from [PX4-Autopilot/msg](https://github.com/PX4/PX4-Autopilot/tree/main/msg) to its `msg` fo ...
```
```suggestion
- If you're creating or modifying uORB messages you must manually update the messages in your workspace from your PX4 source tree.
  Generally this means that you would update [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml), clone the interface package, and then manually synchronize it by copying the new/modified message definitions from [PX4-Autopilot/msg](https://github.com/PX4/PX4-Autopilot/tree/main/msg) to its `msg` folders.
  Assuming that PX4-Autopilot is in your home directory `~`, while `px4_msgs` is in `~/px4_ros_com/src/`, then the command might be:

```

beniaminopozzan commented on en/ros/ros2_comm.md:
```diff
@@ -448,23 +489,51 @@ For more information see [Starting XRCE-DDS](../middleware/xrce_dds.md#starting-
+
+Custom topic namespaces can be applied at build time (changing [dds_topics.yaml](https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/microdds_client/dds_topics.yaml)) or at runtime (useful for multi vehicle operations):
+
+ - One possibility is to use the `-n` option when starting the [microdds-client](../modules/modules_system.md#microdds-client) from command line. This technique can be used both in simulation and real vehicles.
+ - Only in simulation, where the client is automatically started, a custom namespace can be provided setting the environment variable `PX4_MICRODDS_NS` before starting the simulation.
```
```suggestion
 - One possibility is to use the `-n` option when starting the [microdds-client](../modules/modules_system.md#microdds-client) from command line.
   This technique can be used both in simulation and real vehicles.
 - A custom namespace can be provided for simulations (only) by setting the environment variable `PX4_MICRODDS_NS` before starting the simulation.
```

beniaminopozzan commented on en/middleware/xrce_dds.md:
```diff
@@ -257,24 +260,59 @@ The topics are release specific (support is compiled into [microdds-client](../m
+
+ - One possibility is to use the `-n` option when starting the [microdds-client](../modules/modules_system.md#microdds-client) from command line.
+   This technique can be used both in simulation and real vehicles.
+ - a custom namespace can be provided for simulations (only) by setting the environment variable `PX4_MICRODDS_NS` before starting the simulation.
```
```suggestion
 - A custom namespace can be provided for simulations (only) by setting the environment variable `PX4_MICRODDS_NS` before starting the simulation.
```

beniaminopozzan commented:
> @beniaminopozzan OK, I moved both headings up one level, and added this heading here for DDS Topics YAML. I have linked to this section (which duplicates your original text) from above and from the ROS2 Guide). We might even have more information/links here, but this is enough for now - i.e. in some places we talk a little about the file format where we could redirect here.
> 
> Before I did all that I rebased on main, so this now includes all your other PRs.
> 
> I'm going to leave with you to think about this as I have to go catch a plane. If you want to do a further round of edits let me know, otherwise we can merge and add next bits as a post-process. UP to you. I'll look in to see if you have commented in a couple of hours.

Thanks @hamishwillee !

I fixed a few remaining typos and mirrored your changes to _Supported uORB Messages_ and _Customizing the Topic Namespace_ from the XRCE-DDS page to ros_com.

So, now it only needs PX4/PX4-Autopilot#21375 to be merged and then it is good to!

hamishwillee reviewed (approved):
Thank you. Looks good. Feel free to merge when that goes in!!!

hamishwillee commented:
Note, it doesn't have to happen for v1.14 but this is a reminder to update the docs that talk about ROS frames to use/link to the new master doc.

Pedro-Roque commented:
@hamishwillee I fixed the conflicts, please check if it is the fix is good. The associated PR was merged by @dagar

hamishwillee commented:
Thanks @Pedro-Roque !
@beniaminopozzan COngrats on getting the PX4-Autopilot fix in!!!! Merging this now. 

Next steps, I believe will be to start tidying up frame convention stuff in other docs, by pointing them to the stuff you made. No urgency, since those are clearly ROS 1 docs.
